coinstacks/bitcoin/api/src/websocket.ts

```diff
@@ -0,0 +1,83 @@
+import WebSocket from 'isomorphic-ws'
+import { ErrorResponse, RequestPayload, Topics, TxsTopicData } from '@shapeshiftoss/common-api'
+import { SequencedBTCParseTx } from '@shapeshiftoss/bitcoin-ingester'
+
+export interface Connection {
+  ws: WebSocket
+  pingTimeout?: NodeJS.Timeout
+}
+
+export class Client {
+  private readonly url: string
+  private readonly connections: Record<Topics, Connection | undefined>
+
+  constructor(url: string) {
+    this.url = url
+
+    this.connections = {
+      txs: undefined,
+    }
+  }
+
+  private heartbeat(topic: Topics): void {
+    const connection = this.connections[topic]
+    if (!connection) return
+
+    connection.pingTimeout && clearTimeout(connection.pingTimeout)
+    connection.pingTimeout = setTimeout(() => connection?.ws.close(), 10000 + 1000)
+  }
+
+  private onOpen(topic: Topics, resolve: (value: unknown) => void): void {
+    this.heartbeat(topic)
+    resolve(true)
+  }
+
+  async subscribeTxs(
+    data: TxsTopicData,
+    onMessage: (message: SequencedBTCParseTx) => void,
+    onError?: (err: ErrorResponse) => void
+  ): Promise<void> {
+    if (this.connections.txs) return
+
+    const ws = new WebSocket(this.url)
+    this.connections.txs = { ws }
+
+    if (onError) {
+      ws.onerror = (event) => onError({ type: 'error', message: event.message })
+    }
+
+    ws.onclose = () => this.connections.txs?.pingTimeout && clearTimeout(this.connections.txs.pingTimeout)
+    ws.onmessage = (event) => {
+      if (event.data === 'ping') {
+        this.heartbeat('txs')
+        return
+      }
+
+      try {
+        const message = JSON.parse(event.data.toString()) as SequencedBTCParseTx | ErrorResponse
+        if ('type' in message) {
```

Comment:
Can you add a comment explaining why you're checking for `type` here?

Reply:
sure!